### PR TITLE
Place 2.0 change from enum to struct

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 1]
```diff
@@ -615,27 +624,47 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                    base,
+                    projection,
+                } = first_borrowed_place;
+
+                let mut current = projection;
+
+                while let Some(box PlaceProjection { base: base_proj, elem }) = current {
                     match elem {
-                        ProjectionElem::Field(field, _) if is_union(base) => {
+                        ProjectionElem::Field(field, _) if is_union(&Place {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/error_reporting.rs]: change `is_union` to take two arguments that are forwareded to `Place::ty_from` instead of creating a place just to call `Place::ty`

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 2]
```diff
@@ -644,11 +673,11 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
 
                             // Also compute the name of the union type, eg. `Foo` so we
                             // can add a helpful note with it.
-                            let ty = base.ty(self.mir, self.infcx.tcx).ty;
+                            let ty = Place::ty_from(&base, &proj_base, self.mir, self.infcx.tcx).ty;
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/error_reporting.rs]: add a FIXME to ensure we don't compute the type twice (and instead have `is_union` return it)

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 3]
```diff
@@ -1658,44 +1727,66 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                         } else {
                             if autoderef {
                                 self.append_place_to_string(
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/error_reporting.rs]: leave a FIXME to turn this recursion into iteration

File: src/librustc_codegen_ssa/mir/place.rs [hunk 4]
```diff
@@ -410,17 +413,22 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
-            mir::Place::Base(
-                mir::PlaceBase::Static(
-                    box mir::Static { ty, kind: mir::StaticKind::Promoted(promoted) }
-                )
-            ) => {
+        let result = match place {
+            mir::Place {
+                base: mir::PlaceBase::Local(_),
+                projection: None,
+            } => bug!(), // handled above
```
Comment (oli-obk) [inline src/librustc_codegen_ssa/mir/place.rs]: preexisting: move the code from above to here.

File: src/librustc_codegen_ssa/mir/place.rs [hunk 5]
```diff
@@ -443,26 +451,40 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
+                    projection: proj_base.clone(),
+                }).deref(bx.cx())
             }
-            mir::Place::Projection(ref projection) => {
-                let cg_base = self.codegen_place(bx, &projection.base);
+            mir::Place {
+                base,
+                projection: Some(projection),
+            } => {
+                let cg_base = self.codegen_place(bx, &mir::Place {
```
Comment (oli-obk) [inline src/librustc_codegen_ssa/mir/place.rs]: leave a FIXME to turn this recursion into iteration

File: src/librustc_codegen_ssa/mir/operand.rs [hunk 6]
```diff
@@ -397,8 +400,14 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
 
         // Moves out of scalar and scalar pair fields are trivial.
-        if let &mir::Place::Projection(ref proj) = place {
-            if let Some(o) = self.maybe_codegen_consume_direct(bx, &proj.base) {
+        if let mir::Place {
+            base,
+            projection: Some(proj),
+        } = place {
+            if let Some(o) = self.maybe_codegen_consume_direct(bx, &mir::Place {
```
Comment (oli-obk) [inline src/librustc_codegen_ssa/mir/operand.rs]: FIXME: iterate instead of recurse

**Comment by spastorino** [comment]
I've just added some commits that are right now PRs waiting on bors and rebased this work on top of them, because this work need those. Once they are merged I'm going to rebase this PR on top of master again. Anyway, the important bit here is the [WIP commit](https://github.com/rust-lang/rust/pull/60913/commits/0c4ed32fe6de74efad4318bfd1b2c8606606623d)

File: src/librustc/mir/visit.rs [hunk 7]
```diff
@@ -675,20 +675,12 @@ macro_rules! make_mir_visitor {
...
-                    }
-                    Place::Projection(proj) => {
-                        let context = if context.is_mutating_use() {
-                            PlaceContext::MutatingUse(MutatingUseContext::Projection)
-                        } else {
-                            PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
-                        };
+                let Place { base, projection } = place;
 
-                        self.visit_place(& $($mutability)? proj.base, context, location);
```
Comment (oli-obk) [inline src/librustc/mir/visit.rs]: just sanity checking: have you gone through all `visit_place` implementations in rustc and checked whether they should manually iterate?

File: src/librustc/mir/visit.rs [hunk 8]
```diff
@@ -675,20 +675,12 @@ macro_rules! make_mir_visitor {
...
-                        self.visit_projection(proj, location);
-                    }
+                self.visit_place_base(base, context, location);
+
+                if let Some(proj) = projection {
+                    self.visit_projection(proj, location);
```
Comment (oli-obk) [inline src/librustc/mir/visit.rs]: the old code recursively visited all projections. I think you should use iterate here in order to visit all of them, too.

This may in fact be what causes the current ICE, some types aren't getting visited.

File: src/librustc_codegen_ssa/mir/analyze.rs [hunk 9]
```diff
@@ -192,7 +195,10 @@ impl<'mir, 'a: 'mir, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
                 // A deref projection only reads the pointer, never needs the place.
                 if let mir::ProjectionElem::Deref = proj.elem {
                     return self.visit_place(
-                        &proj.base,
+                        &mir::Place {
+                            base: place.base.clone(),
```
Comment (oli-obk) [inline src/librustc_codegen_ssa/mir/analyze.rs]: leave a FIXME here

File: src/librustc_mir/borrow_check/mod.rs [hunk 10]
```diff
@@ -1698,18 +1713,26 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     fn move_path_closest_to<'a>(
         &mut self,
         place: &'a Place<'tcx>,
-    ) -> Result<(&'a Place<'tcx>, MovePathIndex), NoMovePathFound> where 'cx: 'a {
-        let mut last_prefix = place;
-        for prefix in self.prefixes(place, PrefixSet::All) {
-            if let Some(mpi) = self.move_path_for_place(prefix) {
-                return Ok((prefix, mpi));
+    ) -> Result<(Place<'tcx>, MovePathIndex), NoMovePathFound> where 'cx: 'a {
+        let mut last_prefix = place.base.clone();
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mod.rs]: no need to clone, just do `&place.base`

File: src/librustc_mir/borrow_check/mod.rs [hunk 11]
```diff
@@ -1698,18 +1713,26 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                projection: prefix.1.clone(),
+            }) {
+                return Ok((Place {
+                    base: prefix.0.clone(),
+                    projection: prefix.1.clone(),
+                }, mpi));
             }
-            last_prefix = prefix;
+
+            last_prefix = prefix.0.clone();
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mod.rs]: and make this `last_prefix = prefix.0`;

File: src/librustc_mir/borrow_check/mod.rs [hunk 12]
```diff
@@ -1698,18 +1713,26 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
-    ) -> Result<(&'a Place<'tcx>, MovePathIndex), NoMovePathFound> where 'cx: 'a {
-        let mut last_prefix = place;
-        for prefix in self.prefixes(place, PrefixSet::All) {
-            if let Some(mpi) = self.move_path_for_place(prefix) {
-                return Ok((prefix, mpi));
+    ) -> Result<(Place<'tcx>, MovePathIndex), NoMovePathFound> where 'cx: 'a {
+        let mut last_prefix = place.base.clone();
+
+        for prefix in self.prefixes(&place.base, &place.projection, PrefixSet::All) {
+            if let Some(mpi) = self.move_path_for_place(&Place {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mod.rs]: can you make `move_path_for_place` clone-free, by giving it two args?

**Review by oli-obk — CHANGES_REQUESTED** [review]
preliminary review

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 13]
```diff
@@ -339,21 +342,22 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
+        while let Some(box Projection { base: Some(ref proj), .. }) = base_static {
+            base_static = &proj.base;
         }
+        let base_static = Place {
+            base: place.base.clone(),
+            projection: base_static.clone(),
+        };
 
-        let description = if let Place::Base(_) = place {
+        let description = if place.projection == None {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/move_errors.rs]:
```suggestion
        let description = if place.projection.is_none() {
```

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 14]
```diff
@@ -339,21 +342,22 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
             format!("static item `{}`", self.describe_place(place).unwrap())
         } else {
             format!(
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/move_errors.rs]: create the base_static in the `else` only

File: src/librustc/mir/visit.rs [hunk 15]
```diff
@@ -675,20 +675,12 @@ macro_rules! make_mir_visitor {
...
-                            PlaceContext::MutatingUse(MutatingUseContext::Projection)
-                        } else {
-                            PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
-                        };
+                let Place { base, projection } = place;
 
-                        self.visit_place(& $($mutability)? proj.base, context, location);
-                        self.visit_projection(proj, location);
-                    }
+                self.visit_place_base(base, context, location);
```
Comment (oli-obk) [inline src/librustc/mir/visit.rs]: the mutability errors happen because the `context` here has not been mutated through all the projections. I guess pass down the base as an argument to `visit_projection` and once you hit a `None`, call `visit_place_base`

File: src/librustc/mir/visit.rs (hunk 7 above)
Comment (spastorino) [reply src/librustc/mir/visit.rs]: Yes, have done that.

File: src/librustc/mir/visit.rs (hunk 8 above)
Comment (spastorino) [reply src/librustc/mir/visit.rs]: Right, wrong rebase. Done.

File: src/librustc_mir/borrow_check/mod.rs (hunk 12 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/mod.rs]: The idea is to do that in this same PR, next commits. Unsure if you still want me to do that now.

File: src/librustc/mir/visit.rs (hunk 15 above)
Comment (spastorino) [reply src/librustc/mir/visit.rs]: True, that was lost on the rebase. Unsure why do you want that to happen there. We can just add the `context` adjustments to `visit_place` the only problem is that implementors would need to do the same :/.

File: src/librustc_mir/borrow_check/mod.rs (hunk 12 above)
Comment (oli-obk) [reply src/librustc_mir/borrow_check/mod.rs]: nope that's fine to do later

File: src/librustc/mir/visit.rs (hunk 15 above)
Comment (oli-obk) [reply src/librustc/mir/visit.rs]: what would be the alternative? wrap the `context` mutation in an `if projection.is_some()`?

Comment (spastorino) [reply src/librustc/mir/visit.rs]: Yep as you can see https://github.com/rust-lang/rust/pull/60913/commits/2c82f6d817e6aa18a85624cb4bf694c1433c021f

File: src/librustc_mir/transform/generator.rs [hunk 16]
```diff
@@ -130,9 +130,9 @@ impl<'tcx> MutVisitor<'tcx> for PinArgVisitor<'tcx> {
                     place: &mut Place<'tcx>,
                     context: PlaceContext,
                     location: Location) {
-        if *place == Place::Base(PlaceBase::Local(self_arg())) {
-            *place = Place::Projection(Box::new(Projection {
-                base: place.clone(),
+        if *place == Place::from(self_arg()) {
```
Comment (oli-obk) [inline src/librustc_mir/transform/generator.rs]: this (and all others in this file), should be `if place.base == PlaceBase::Local(self_arg())`, but the change should not be applied to the outermost projection, but to the innermost one. You could create a new free function in this file that replaces the innermost `None` projetion with `Some(Box::new(Projection { base: None, elem: argument_to_new_free_func }))`

File: src/librustc_mir/borrow_check/mod.rs [hunk 17]
```diff
@@ -2083,62 +2128,77 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
                     Ok(RootPlace {
-                        place,
+                        place_base,
+                        place_projection,
                         is_local_mutation_allowed,
                     })
                 }
             }
-            Place::Projection(ref proj) => {
+            (ref base, Some(ref proj)) => {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mod.rs]: the mixed use of `base` and `place_base` is a little confusing, just use `place_base` everywhere

```suggestion
            (_, Some(ref proj)) => {
```

File: src/librustc/mir/visit.rs [hunk 18]
```diff
@@ -676,20 +677,17 @@ macro_rules! make_mir_visitor {
...
+                    let context = if context.is_mutating_use() {
+                        PlaceContext::MutatingUse(MutatingUseContext::Projection)
+                    } else {
+                        PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
+                    };
+
+                    self.visit_projection(& $($mutability)? place.base, proj, context, location);
                 }
+
+                self.visit_place_base(& $($mutability)? place.base, context, location);
```
Comment (oli-obk) [inline src/librustc/mir/visit.rs]: if I'm reading this diff right, then `context` here used to be `*UseContext::Projection` if there were any projections.

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 19]
```diff
@@ -591,8 +600,8 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
     ) -> (String, String, String, String) {
         // Define a small closure that we can use to check if the type of a place
         // is a union.
-        let union_ty = |place: &Place<'tcx>| -> Option<Ty<'tcx>> {
-            let ty = place.ty(self.body, self.infcx.tcx).ty;
+        let union_ty = |place_base: &PlaceBase<'tcx>, place_projection: &Option<Box<Projection<'tcx>>>| -> Option<Ty<'tcx>> {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: I don't think these type annotations are necessary. `Place::ty_from` is nongeneric

File: src/librustc/mir/visit.rs [hunk 20]
```diff
@@ -676,20 +677,19 @@ macro_rules! make_mir_visitor {
...
+                    context = if context.is_mutating_use() {
+                        PlaceContext::MutatingUse(MutatingUseContext::Projection)
+                    } else {
+                        PlaceContext::NonMutatingUse(NonMutatingUseContext::Projection)
+                    };
+
+                    self.visit_projection(& $($mutability)? place.base, proj, context, location);
                 }
+
+                self.visit_place_base(& $($mutability)? place.base, context, location);
```
Comment (mark-i-m) [inline src/librustc/mir/visit.rs]: Is it intentional that the `place_base` is visited _after_ the projections with the updated context?

Comment (spastorino) [reply src/librustc/mir/visit.rs]: @mark-i-m yeah, this was due to a rebase problem I think. I've fixed this yesterday but thanks for checking :)

Comment (spastorino) [reply src/librustc/mir/visit.rs]: @mark-i-m yes, this is exactly what the previous code did if you follow the recursion.

File: src/librustc/mir/visit.rs (hunk 18 above)
Comment (eddyb) [reply src/librustc/mir/visit.rs]: I wonder if we can get rid of `UseContext::Projection` altogether, now that we don't need recursion for `visit_place` - but I'd leave this for another PR.

File: src/librustc_mir/transform/promote_consts.rs [hunk 21]
```diff
@@ -308,13 +312,13 @@ impl<'a, 'tcx> Promoter<'a, 'tcx> {
                     match statement.kind {
                         StatementKind::Assign(_, box Rvalue::Ref(_, _, ref mut place)) => {
                             // Find the underlying local for this (necessarily interior) borrow.
-                            let mut place = place;
-                            while let Place::Projection(ref mut proj) = *place {
+                            let mut place_projection = &place.projection;
+                            while let Some(proj) = place_projection {
                                 assert_ne!(proj.elem, ProjectionElem::Deref);
-                                place = &mut proj.base;
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc_mir/transform/promote_consts.rs]: This entire loop can be removed, we're just interested in the base

File: src/librustc_mir/borrow_check/mod.rs [hunk 22]
```diff
@@ -2179,17 +2242,21 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+                        let upvar_field_projection = self.is_upvar_field_projection(&Place {
+                            base: place_base.clone(),
+                            projection: place_projection.clone(),
+                        });
                         if let Some(field) = upvar_field_projection {
                             let upvar = &self.upvars[field.index()];
                             debug!(
-                                "upvar.mutability={:?} local_mutation_is_allowed={:?} place={:?}",
-                                upvar, is_local_mutation_allowed, place
+                                "upvar.mutability={:?} local_mutation_is_allowed={:?}" /
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mod.rs]:
```suggestion
                                "upvar.mutability={:?} local_mutation_is_allowed={:?} \
```

File: src/librustc_mir/borrow_check/mod.rs [hunk 23]
```diff
@@ -2179,17 +2242,21 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+                                "place={:?} {:?}",
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mod.rs]:
```suggestion
                                place={:?} {:?}",
```

**Review by oli-obk — COMMENTED** [review]
this should fix the travis failure

File: src/librustc_mir/borrow_check/mod.rs (hunk 22 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/mod.rs]: Hehe I was sleepy yesterday trying to push something quickly and wrote that in another language :P.

**Comment by oli-obk** [comment]
@rust-timer build 0c7cf8c633b388d0d8cc4991d41a281d6ddd47bd

File: src/librustc/mir/mod.rs [hunk 24]
```diff
@@ -1862,54 +1874,77 @@ impl<'tcx> Place<'tcx> {
...
+                    )
+                }
             }
+        }
 
-            Place::Base(base) => op(base, next.iter()),
+        iterate_over2(place_base, place_projection, &Projections::Empty, op)
+    }
+
+    pub fn as_place_ref(&self) -> PlaceRef<'_, 'tcx> {
```
Comment (oli-obk) [inline src/librustc/mir/mod.rs]: if it is deemed too wordy, I'm also fine with using `as_ref`. We can't use the `AsRef` trait though as that produces a reference

File: src/librustc/mir/mod.rs [hunk 25]
```diff
@@ -1761,7 +1761,7 @@ impl_stable_hash_for!(struct Static<'tcx> {
     Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, RustcEncodable, RustcDecodable, HashStable,
 )]
 pub struct Projection<'tcx> {
-    pub base: Place<'tcx>,
+    pub base: Option<Box<Projection<'tcx>>>,
```
Comment (RalfJung) [inline src/librustc/mir/mod.rs]: It is somewhat confusing to have `base` sometimes be a `Projection` and sometimes a `PlaceBase`.

Comment (oli-obk) [reply src/librustc/mir/mod.rs]: Yes it is, but the entire Projection type will be gone soon, so I deemed it irrelevant
